pandas/_libs/tslibs/c_timestamp.pyx

```diff
@@ -287,9 +287,13 @@ cdef class _Timestamp(datetime):
         if (PyDateTime_Check(self)
                 and (PyDateTime_Check(other) or is_datetime64_object(other))):
             if isinstance(self, _Timestamp):
-                other = type(self)(other)
+                both_timestamps = isinstance(other, _Timestamp)
```

Comment:
@jbrockmendel can you take another look when you get a chance? Needed to change some things around to ensure that `Timestamp - datetime.datetime` didn't raise (since we internally convert the datetime to a Timestamp).

Reply:
this looks like you're avoiding re-calling the constructor if we already have a Timestamp.  Is that necessary?  The constructor was recently optimized for the already-Timestamp case

Reply:
never mind, i get it now

Reply:
maybe add a comment to the effect of "we need to know if we started with both-Timestamps to determine whether to raise on Overflow, see GH#..."